README.md
# VRipsTDA
R-Package for plotting Vietoris–Rips complexes in R


NEWS.md
# VRipsTDA 1.0.0

* Initial CRAN submission.


R/staticVRips.R
#' Create static Vietoris-Rips complexes
#'
#' @description [staticVRips()] takes the input of a dataframe with two columns, calculates the persistence homology, 
# and creates a graph of the complex given a value of tau. This can also show the current barcode
# and create a png of the complex.
#' 
#' @param df A data frame with two columns to create the Vietoris-Rips from.
#' @param tau A value of to compute the Vietoris-Rips for.
#' @param extra_thresh The additional threshold for the function to use when graphing the complex. This adjusts the bounds of the graph.
#' @param inc_barcode If [TRUE], the image includes the persistence barcode plot.
#' @param png If [TRUE], the png of the output is saved in the current directory.
#' @param filename The name of the png file.
#' @param ... Additional parameters that can be passed through `ggsave()`
#'
#' @return A plot of the Vietoris-Rips complex with given value tau and the barcode plot.
#' 
#' @export
#'
staticVRips <- function(df, tau, extra_thresh = 0.25, inc_barcode = FALSE,
                        png = FALSE, filename = "VRComplexStatic_plot",...){
  
  ph <- calculate_homology(df)
  
  ph_df <- as.data.frame(ph)
  ph_barcode_df <- ph_df %>%
    filter(birth <= tau) %>%
    mutate(death = ifelse(death > tau, tau, death))
  ph_barcode_matrix <- as.matrix(ph_barcode_df)
  
  barcode_plot <- plot_barcode(ph_barcode_matrix)
  
  taus_needed <- data.frame(taus_needed_nondf = sort(union(ph_df$death, ph_df$birth)))
  
  # Calculate intersections and initiate triangles df
  
  intersections <- intersection(df, tau)
  triangles <- data.frame()
  
  # Initial plot if there are no segments. Followed by just the segments
  VRplot <- ggplot(data = df, aes(.data[[colnames(df)[1]]], .data[[colnames(df)[2]]]))
    geom_point() +
    geom_circle(aes(x0 = .data[[colnames(df)[1]]], y0 = .data[[colnames(df)[2]]], r = tau), fill = NA, color = "blue") +
    xlim(-max(taus_needed) - 10*extra_thresh, max(taus_needed) + 10*extra_thresh) +
    ylim(-max(taus_needed) - 10*extra_thresh, max(taus_needed) + 10*extra_thresh) +
    coord_quickmap() +
    theme_classic() +
    annotate("text", label = paste("Tau =", tau), 
             x = max(taus_needed) + 10*extra_thresh, y = -max(taus_needed) - 10*extra_thresh, vjust = 0, hjust = 1)
  
  if(nrow(intersections) != 0){
    colnames(intersections) <- c('xstart', 'ystart', 'xend', 'yend')
    VRplot <- VRplot +
      geom_segment(data = intersections, aes(x = xstart, y = ystart, xend = xend, yend = yend), color = "red")
    
    # Make df for triangle calculation
    triangle_test_points <- tibble(x = c(intersections$xstart, intersections$xend), 
                                   y = c(intersections$ystart, intersections$yend)) %>%
      unique()
    
    # Calculate triangles
    triangles <- triangle_calculate(triangle_test_points, tau)
    
  }
  
  if(nrow(triangles) != 0){
    VRplot <- VRplot +
      geom_polygon(data = triangles, aes(x = x, y = y, group = group), color = "red", fill = "lightgreen", alpha = 0.4)
  }
  
  if(png){
    
    if(inc_barcode){
      suppressGraphics(suppressMessages(ggsave(paste(filename, ".png", sep = ""),grid.arrange(VRplot,barcode_plot, ncol = 2))))
    } else{
      suppressGraphics(suppressMessages(ggsave(paste(filename, ".png", sep = ""),VRplot)))
    }
  }
  
  if(inc_barcode){
    grid.arrange(VRplot,barcode_plot, ncol = 2)
  } else{
    return(VRplot)
  }
  
}

R/triangle_calculate.R
#' Find triangles between three points
#'
#' @description [trianglecalculate()] is a function to calculate triangles between three points.
#' This is a helper function for [aniVRips()] and [staticVRips()]
#' 
#' @param df A dataframe with two columns
#' @param r A radial distance to calculate if triangles are formed.
#'
#' @return A dataframe with three columns x, y, and group where group is the indicator
#' of what triangle the points are apart of.
#' @export
#'
triangle_calculate <- function(df, r){
  
  triangles <- data.frame()
  
  colnames(df) <- c('x', 'y')
  
  group_counter <- 1
  groups <- data.frame(x = numeric(0), y = numeric(0), group = numeric(0))
  
  for (i in 1:(nrow(df))) {
    for (j in (i):(nrow(df))) {
      for (k in (j):nrow(df)) {
        if (i != j && j != k && k != i){
          d1 <- distance(df$x[i], df$y[i], df$x[j], df$y[j])
          d2 <- distance(df$x[i], df$y[i], df$x[k], df$y[k])
          d3 <- distance(df$x[j], df$y[j], df$x[k], df$y[k])
          
          if (d1 <= 2*r && d2 <= 2*r && d3 <= 2*r) {
            group <- data.frame(x = c(df$x[i], df$x[j], df$x[k]),
                                y = c(df$y[i], df$y[j], df$y[k]),
                                group = group_counter)
            groups <- rbind(groups, group)
            group_counter <- group_counter + 1
          }
        }
      }
    }
    
  }
  return(groups)
}

R/aniVRips.R
#' Create animated Vietoris-Rips complexes
#' @description [aniVRips()] takes the input of a dataframe with two columns, 
#' calculates the persistence homology, and creates an HTML animation showing 
#' the formation of the complex over tau.
#' 
#' @param df A data frame with two columns to create the Vietoris-Rips from
#' @param framerate The number of frames you want the animation to use
#' @param extra_thresh The additional threshold for the function to use when graphing the complex. This adjusts the bounds of the graph.
#' @param inc_barcode If [TRUE], the animation includes the persistence barcode plot.
#' @param filename The file name of the .html file.
#' @param title The title of the .html file.
#' @param description The description of the animation.
#'
#' @return An HTML file and its corresponding plots.
#' @export
#'
#' 
#'
aniVRips <- function(df, framerate, extra_thresh = 0.25, inc_barcode = FALSE, filename = "VRComplexAni",
                     title = "Animation of V-R Complex over Tau", 
                     description = "This animation represents the creation of the V-R complex of some set \n of data. Very cool!"){
  ph <- calculate_homology(df)
  
  ph_df <- as.data.frame(ph)
  
  taus_needed <- data.frame(taus_needed_nondf = sort(union(ph_df$death, ph_df$birth)))
  
  tau_seq <- seq(from = min(taus_needed), to = max(taus_needed) + extra_thresh, length.out = framerate)
  
  saveHTML({
    for (tau in tau_seq){
      
      # Making the barcode for the given tau
      ph_barcode_df <- ph_df %>%
        filter(birth <= tau) %>%
        mutate(death = ifelse(death > tau, tau, death))
      ph_barcode_matrix <- as.matrix(ph_barcode_df)
      barcode_plot <- plot_barcode(ph_barcode_matrix)
      
      # Calculate intersections and initiate triangles df
      
      intersections <- intersection(df, tau)
      triangles <- data.frame()
      
      # Initial plot if there are no segments. Followed by just the segments
      VRplot <- ggplot(data = df, aes(.data[[colnames(df)[1]]], .data[[colnames(df)[2]]]))
        geom_point() +
        geom_circle(aes(x0 = .data[[colnames(df)[1]]], y0 = .data[[colnames(df)[2]]], r = tau), fill = NA, color = "blue") +
        xlim(-max(taus_needed) - 10*extra_thresh, max(taus_needed) + 10*extra_thresh) +
        ylim(-max(taus_needed) - 10*extra_thresh, max(taus_needed) + 10*extra_thresh) +
        coord_quickmap() +
        theme_classic() +
        annotate("text", label = paste("Tau =", tau), 
                 x = max(taus_needed) + 10*extra_thresh, y = -max(taus_needed) - 10*extra_thresh, vjust = 0, hjust = 1)
      
      if(nrow(intersections) != 0){
        colnames(intersections) <- c('xstart', 'ystart', 'xend', 'yend')
        VRplot <- VRplot +
          geom_segment(data = intersections, aes(x = xstart, y = ystart, xend = xend, yend = yend), color = "red")
        
        # Make df for triangle calculation
        triangle_test_points <- tibble(x = c(intersections$xstart, intersections$xend), 
                                       y = c(intersections$ystart, intersections$yend)) %>%
          unique()
        
        # Calculate triangles
        triangles <- triangle_calculate(triangle_test_points, tau)
        
      }
      
      if(nrow(triangles) != 0){
        VRplot <- VRplot +
          geom_polygon(data = triangles, aes(x = x, y = y, group = group), color = "red", fill = "lightgreen", alpha = 0.4)
      }
      if(inc_barcode){
        grid.arrange(VRplot,barcode_plot, ncol = 2)
      } else{
        print(VRplot)
      }
      ani.pause()
    }
  },
  img.name = paste(filename, "_plot", sep = ""), imgdir = paste(filename, "_dir", sep = ""), htmlfile = paste(filename, ".html", sep = ""), 
  autobrowse = FALSE, title = title, 
  description = description)
}